src/FormRequest.php

```diff
@@ -0,0 +1,61 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Form;
+
+use Psr\Http\Message\ServerRequestInterface;
+use Yiisoft\Validator\ValidatorInterface;
+
+/**
+ * The abstraction for common case when using a form in a controller
+ */
+abstract class FormRequest
+{
+    private ?string $formName = null;
+    private bool $isHandled = false;
+
+    public function __construct(
+        protected ServerRequestInterface $serverRequest,
+        protected ValidatorInterface $validator,
+        protected FormModel $formModel
+    ) {
+    }
+
+    public function isValid(): bool
+    {
+        if (!$this->isHandled) {
+            $this->handle();
+        }
+        return !$this->formModel()->getFormErrors()->hasErrors();
+    }
+
+    public function formModel(): FormModel
```

Comment:
The method makes two things at once: handles the form and returns the model. I'd try to avoid it and do these two things separately to avoid side effects.